src/groupeddataframe/grouping.jl

```diff
@@ -181,6 +197,319 @@ end
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 2     │ 7     │
+⋮
+Last Group (2 rows): a = :baz, b = 1
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ baz    │ 1     │ 6     │
+│ 2   │ baz    │ 1     │ 12    │
+
+julia> collect(groupvalues(gd))
+6-element Array{NamedTuple{(:a, :b),Tuple{Symbol,Int64}},1}:
+ (a = :foo, b = 2)
+ (a = :bar, b = 1)
+ (a = :baz, b = 2)
+ (a = :foo, b = 1)
+ (a = :bar, b = 2)
+ (a = :baz, b = 1)
+
+julia> groupvalues(gd, 2)
+(a = :bar, b = 1)
+
+julia> groupvalues(gd, 2, 1)
+:bar
+
+julia> groupvalues(gd, 2, :a)
+:bar
+
+julia> groupvalues(Tuple, gd, 2)
+(:bar, 1)
+```
+"""
+groupvalues(T::Type, gd::GroupedDataFrame) = (groupvalues(T, gd, i) for i in 1:length(gd))
+groupvalues(gd::GroupedDataFrame) = groupvalues(NamedTuple, gd)
+
+groupvalues(::Type{Tuple}, gd::GroupedDataFrame, i::Integer) = Tuple(gd.parent[gd.idx[gd.starts[i]], gd.cols])
+groupvalues(::Type{NamedTuple}, gd::GroupedDataFrame, i::Integer) = NamedTuple{Tuple(groupvars(gd))}(groupvalues(Tuple, gd, i))
+groupvalues(gd::GroupedDataFrame, i::Integer) = groupvalues(NamedTuple, gd, i)
+
+groupvalues(gd::GroupedDataFrame, i::Integer, col::Integer) = gd.parent[gd.idx[gd.starts[i]], gd.cols[col]]
+groupvalues(gd::GroupedDataFrame, i::Integer, col::Symbol) = groupvalues(gd, i, _groupvar_idx(gd, col))
+
+# Eltypes of the grouping columns
+_grouptypes(gd::GroupedDataFrame) = Tuple(eltype(gd.parent[!, c]) for c in gd.cols)
+
+
+"""
+    GroupKey
+
+Key for one of the groups of a [`GroupedDataFrame`](@ref). Contains the values
+of the corresponding grouping columns and behaves similarly to a `NamedTuple`,
+but using it to index its `GroupedDataFrame` is much more effecient than using the
+equivalent `Tuple` or `NamedTuple`.
+
+Instances of this type are returned by `keys(::GroupedDataFrame)` and are not
+meant to be constructed directly.
+
+See [`keys(::GroupedDataFrame)`](@ref) for more information.
+"""
+struct GroupKey{P<:GroupedDataFrame}
+    parent::P
+    idx::Int
+end
+
+function Base.show(io::IO, k::GroupKey)
+    print(io, "GroupKey")
+    show(io, NamedTuple(k))
+end
+
+Base.parent(key::GroupKey) = getfield(key, :parent)
+Base.length(key::GroupKey) = length(parent(key).cols)
+Base.names(key::GroupKey) = Tuple(groupvars(parent(key)))
+Base.keys(key::GroupKey) = names(key)
+Base.propertynames(::GroupKey) = names(key)
+Base.values(key::GroupKey) = groupvalues(Tuple, parent(key), getfield(key, :idx))
+
+Base.iterate(key::GroupKey) = iterate(key, 1)
+Base.iterate(key::GroupKey, i::Integer) = i <= length(key) ? (key[i], i + 1) : nothing
+
+Base.getindex(key::GroupKey, i::Integer) = groupvalues(parent(key), getfield(key, :idx), i)
+
+function Base.getindex(key::GroupKey, n::Symbol)
+    try
+        return groupvalues(parent(key), getfield(key, :idx), n)
+    catch e
+        throw(KeyError(n))
+    end
+end
+
+function Base.getproperty(key::GroupKey, p::Symbol)
+    try
+        return key[p]
```

Comment:
Similarly here - better make `key[p]` throw an appropriate error I think (and again - we avoid using `error` as it is too generic).

Reply:
I'm using `ErrorException` here to match the default behavior on an attempt to access a non-existent field, which happens with `NamedTuple`.